app/javascript/components/UI/rich-text/rich-text.js

```diff
@@ -0,0 +1,165 @@
+    return obj( str.slice(0, cursorPosition-n) + str.slice(cursorPosition) )
+
+  }
+  static pasteCharsToString( chars, str,  startPos, endPos = startPos) {
+    return str.slice(0, startPos) + chars + str.slice(endPos)
+  }
+  static  pasteSymbolsToHTMLobj(symbols, htmlText, cursorPos, setObjHTML, setCaret) {
+    setObjHTML(this.encodeSpace(htmlText.slice(0, cursorPos.realPos) + this.encodeSpace(symbols)
+        + htmlText.slice(cursorPos.realPos)))
+    setCaret(cursorPos.charCount  + symbols.length)
+  }
+
+  static  pasteCharsBeforeEndTag(chars, htmlText, cursorPos, endTag, setObjHTML, setCaret) {
+    setObjHTML(this.encodeSpace(htmlText.slice(0, cursorPos.realPos) + this.encodeSpace(chars) +
+        htmlText.slice(htmlText.indexOf(endTag, cursorPos.realPos))))
+    setCaret(cursorPos.charCount  + chars.length)
+  }
+
+  static  pasteNodeToHTMLobj(symbols, htmlText, cursorPos, setObjHTML, setCaret, tag, endTag) {
+    setObjHTML(this.encodeSpace(htmlText.slice(0, cursorPos.realPos) + tag +
+        this.encodeSpace(symbols) + endTag + htmlText.slice(cursorPos.realPos)))
+    setCaret(cursorPos.charCount  + symbols.length)
+  }
+
+  static deleteString( string, startPos, endPos ) {
+    return string.slice(0, startPos) + string.slice(endPos)
+  }
+
+  static incrementPositionCursor (numChars, cursorPos, text, obj = ()=>{}) {
+    obj( cursorPos.charCount + numChars < text.length ? cursorPos.charCount + numChars : text.length )
+  }
+  
+  static decrementPositionCursor (numChars, cursorPos, obj = ()=>{}) {
+    obj( cursorPos.charCount > numChars  ? cursorPos.charCount - numChars : 0 )
+  }
+
+  static deleteNode( htmlStr, cursorPos, tag, endTag, setObjHTML, setCaret){
+    const startPos =
+        cursorPos.isSPAN ? cursorPos.realPos - cursorPos.realFocusOffset  - tag.length + 1 : cursorPos.realPos
+    const endPos  = htmlStr.indexOf(endTag, cursorPos.realPos ) + endTag.length
+    console.log(startPos, endPos, this.deleteString(htmlStr, startPos, endPos) , cursorPos.focusOffset , cursorPos)
+    setObjHTML(this.deleteString(htmlStr, startPos, endPos))
+    setCaret(cursorPos.isSPAN ? cursorPos.charCount - cursorPos.focusOffset : cursorPos.charCount)
+    return htmlStr.slice(startPos, endPos)
+  }
+
+  static deleteNodePasteChars( htmlStr, cursorPos, chars, tag, endTag, setObjHTML, setCaret){
+    const startPos = cursorPos.realPos - cursorPos.realFocusOffset  - tag.length + 1
+    const endPos  = htmlStr.indexOf(endTag, cursorPos.realPos ) + endTag.length
+    let newHtmlTxt = this.deleteString(htmlStr, startPos, endPos)
+    setObjHTML(this.pasteCharsToString( chars, newHtmlTxt, startPos))
+    this.incrementPositionCursor( chars.length -1 , cursorPos, newHtmlTxt, setCaret)
+  }
+
+  static findUsersInText( tag, endTag, richText, listUsers) {
+      const users = []
+      let pos = 0
+      while ((pos = richText.indexOf(tag, pos)) !== -1) {
+        pos += + tag.length
```

Comment:
are you sure **+= +** is right operation call here?

Reply:
LOL)